pkg/clusterversion/clusterversion.go

```diff
@@ -90,3 +98,248 @@ func (builder *Builder) validate() (bool, error) {
+			}
+
+			for _, condition := range builder.Object.Status.Conditions {
+				if condition.Type == conditionType {
+					return condition.Status == isTrue, nil
+				}
+			}
+
+			return false, err
+		})
+}
+
+// WaitUntilUpdateIsStarted waits until there is a history entry indicating the update start.
+func (builder *Builder) WaitUntilUpdateIsStarted(timeout time.Duration) error {
+	return builder.WaitUntilUpdateHistoryStateTrue(updateHistoryPartialState, timeout)
+}
+
+// WaitUntilUpdateIsCompleted waits until there is a history entry indicating the update completed.
+func (builder *Builder) WaitUntilUpdateIsCompleted(timeout time.Duration) error {
+	return builder.WaitUntilUpdateHistoryStateTrue(updateHistoryCompletedState, timeout)
+}
+
+// WaitUntilUpdateHistoryStateTrue waits until there is a history entry indicating an updateHistoryState.
+func (builder *Builder) WaitUntilUpdateHistoryStateTrue(
+	updateHistoryState v1.UpdateState, timeout time.Duration) error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	if !builder.Exists() {
+		return fmt.Errorf("%s clusterversion not found", builder.Definition.Name)
+	}
+
+	return wait.PollUntilContextTimeout(
+		context.TODO(), time.Second, timeout, true, func(ctx context.Context) (bool, error) {
+			var err error
+			builder.Object, err = builder.Get()
+
+			if err != nil {
+				return false, nil
+			}
+
+			updateImage := builder.Object.Status.Desired.Image
+
+			for _, updateHistory := range builder.Object.Status.History {
+				if updateHistory.Image == updateImage && updateHistory.State == updateHistoryState {
+					return true, nil
+				}
+			}
+
+			return false, err
+		})
+}
+
+// GetNextUpdateVersionImage fetches the next recommended or conditional update for the cluster.
+func (builder *Builder) GetNextUpdateVersionImage(stream string, acceptConditionalVersions bool) (string, error) {
+	if valid, err := builder.validate(); !valid {
+		return "", err
+	}
+
+	if !builder.Exists() {
+		return "", fmt.Errorf("%s clusterversion not found", builder.Definition.Name)
+	}
+
+	currentVersion := builder.Object.Status.Desired.Version
+
+	for _, availableUpdate := range builder.Object.Status.AvailableUpdates {
+		isStreamUpdate, err := IsStreamUpdate(currentVersion, availableUpdate.Version, stream)
+		if isStreamUpdate && err == nil {
+			return availableUpdate.Image, nil
+		}
+	}
+
+	if acceptConditionalVersions {
+		for _, conditionalUpdate := range builder.Object.Status.ConditionalUpdates {
+			isStreamUpdate, err := IsStreamUpdate(currentVersion, conditionalUpdate.Release.Version, stream)
+			if isStreamUpdate && err == nil {
+				return conditionalUpdate.Release.Image, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Update version in %s stream not found", stream)
+}
+
+// IsStreamUpdate checks if updateVersion is a 'stream' (X, Y or Z) update for version.
+func IsStreamUpdate(version string, updateVersion string, stream string) (isStreamUpdate bool, err error) {
```

Comment:
You can simply define agr type only once:
 
`func IsStreamUpdate(version, updateVersion, stream string) (isStreamUpdate bool, err error) {`

Reply:
Done, thanks